nrf_security: Doc updates for mbed TLS 3.0.0 (#652)
nrfconnect/sdk-nrfxlib: nrf_security/doc/driver_config.rst

.. contents::
   :local:
   :depth: 2

This section covers the configurations available when using PSA drivers.

.. _nrf_security_drivers_config_multiple:

Configuring multiple drivers
****************************

Multiple PSA drivers can be enabled at the same time, with added support for fine-grained control of which drivers implements support for cryptographic features.

To enable a PSA driver, set the configurations according to the following table:

+---------------+------------------------------------+------------------------------------------------+
| PSA driver    | Configurations                     | Notes                                          |
+================+===================================+================================================+
| nrf_cc3xx     | :kconfig:`PSA_CRYPTO_DRIVER_CC3XX` | Only on nRF52840, nRF9160, and nRF5340 devices |
+---------------+------------------------------------+------------------------------------------------+
| nrf_oberon    | :kconfig:`PSA_RYPTO_DRIVER_OBERON` |                                                |
+---------------+------------------------------------+------------------------------------------------+

> If multiple drivers are enabled, the first ordered item in this table takes precedence for an enabled cryptographic feature, unless the driver does not enable or support it.

Reviewer: ```suggestion
When both drivers are enabled the nrf_cc3xx takes precedence for the cryptographic features that it has enabled.
```

We were being overly generic.

Reviewer: discussed offline.

Enabling or disabling PSA driver specific configurations controls the support for a given algorithm, per driver.

> If a specific cryptographic feature is not supported by a PSA driver but the algorithm is configured to be used, then builtin Mbed TLS will be enabled to ensure the feature is available.

Reviewer: "builtin Mbed TLS" needs to be defined.

It should be possible to understand that this is a Nordic fork of mbedtls that is supplied through west.

Reviewer: We are inconsistent. Other places we call it "Orignal mbedtls"

Author: Fixed with more links to the description of what builtin means

AES cipher configuration
************************

AES cipher modes can be enabled by setting one or more of the following Kconfig variables:

+----------------+-----------------------------------------------+
| Cipher mode    | Configurations                                |
+================+===============================================+
| ECB_NO_PADDING | :kconfig:`CONFIG_PSA_WANT_ALG_ECB_NO_PADDING` |
+----------------+-----------------------------------------------+

> | CBC_NO_PADDING | :kconfig:`CONFIG_PSA_WANT_ALG_CTR`            |

Reviewer: ```suggestion
| CBC_NO_PADDING | :kconfig:`CONFIG_PSA_WANT_CBC_NO_PADDING`            |
```

Author: Fixed

Reviewer: Not fixed

| CFB            | :kconfig:`PSA_CRYPTO_DRIVER_ALG_CFB_CC3XX`            | Not supported                                          |
+----------------+-------------------------------------------------------+--------------------------------------------------------+
| CTR            | :kconfig:`PSA_CRYPTO_DRIVER_ALG_CTR_CC3XX`            | :kconfig:`PSA_CRYPTO_DRIVER_ALG_CTR_OBERON`            |
+----------------+-------------------------------------------------------+--------------------------------------------------------+
| OFB            | Not supported                                         | Not supported                                          |
+----------------+-------------------------------------------------------+--------------------------------------------------------+
| XTS            | Not supported                                         | Not supported                                          |
+----------------+-------------------------------------------------------+--------------------------------------------------------+

.. note::
   * If an AES cipher mode is enabled and no PSA driver enables or supports it, then builtin Mbed TLS support is enabled and used.

>    * The :ref:`nrf_security_drivers_cc3xx` is limited to AES key sizes of 128 bits on devices with Arm CryptoCell cc310.

Reviewer: Maybe also mention that otherwise there are two more key sizes supported, unless that is documented elsewhere.

Reviewer: The same applies elsewhere

Author: Not handled in this PR, to be considered for later improvements

+----------------+-------------------------------

MAC driver configurations
=========================

The Kconfig variables in the following table can be used for fine-grained control over which drivers provides AEAD support:

+----------------+---------------------------------------------+----------------------------+
| MAC cipher     | nrf_cc3xx driver support                    | nrf_oberon driver support  |
+================+=============================================+============================+
| ECB_NO_PADDING | :kconfig:`PSA_CRYPTO_DRIVER_ALG_CMAC_CC3XX` | Not supported              |
+----------------+----------------+----------------------------+----------------------------+
| CBC_NO_PADDING | :kconfig:`PSA_CRYPTO_DRIVER_ALG_HMAC_CC3XX` | Not supported              |
+----------------+----------------+----------------------------+----------------------------+

.. note::
   * If a MAC algorithm is enabled and no PSA driver enables or supports it, then builtin Mbed TLS support is enabled and used.
   * The :ref:`nrf_security_drivers_cc3xx` is limited to AES CMAC key sizes of 128 bits on devices with Arm CryptoCell cc310.

>    * The :ref:`nrf_security_drivers_cc3xx` is limited to HMAC using SHA-1, SHA-224, and SHA-256 on devices with Arm CryptoCell cc310.

Reviewer: which hashing algorithms are available outside of cc310 ? We should mention both what is supported and what is not supported.

| ECDH                  | :kconfig:`PSA_WANT_ALG_ECDH`                |
+----------------+----------------------------------------------------+
| ECDSA                 | :kconfig:`PSA_WANT_ALG_ECDSA`               |
+-----------------------+---------------------------------------------+
| ECDSA (deterministic) | :kconfig:`PSA_WANT_ALG_DETERMINISTIC_ECDSA` |
+-----------------------+---------------------------------------------+

The ECC algorithm support is dependent on one or more Kconfig variables enabling curve support according to :ref:`ECC curve configurations`.

ECC driver configurations
=========================

The Kconfig variables in the following table can be used for fine-grained control over which drivers provides ECC support:

+-----------------------+------------------------------------------------------------+-------------------------------------------------------------+
| ECC algorithm         | nrf_cc3xx driver support                                   | nrf_oberon driver support                                   |
+=======================+============================================================+=============================================================+

> | ECDH                  | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDH_CC3XX`                | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDSA`                      |

Reviewer: ```suggestion
| ECDH                  | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDH_CC3XX`                | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDSH_OBERON`                      |
```

Author: Fixed, part of next force-push

+----------------+-------------------------------------------------------------------+-------------------------------------------------------------+
| ECDSA                 | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDSA_CC3XX`               | :kconfig:`PSA_CRYPTO_DRIVER_ALG_ECDSA_OBERON`               |
+-----------------------+------------------------------------------------------------+-------------------------------------------------------------+
| ECDSA (deterministic) | :kconfig:`PSA_CRYPTO_DRIVER_ALG_DETERMINISTIC_ECDSA_CC3XX` | :kconfig:`PSA_CRYPTO_DRIVER_ALG_DETERMINISTIC_ECDSA_OBERON` |
+-----------------------+------------------------------------------------------------+-------------------------------------------------------------+

.. note::
   * If an ECC algorithm is enabled and no PSA driver enables or supports it, then builtin Mbed TLS support is enabled and used.

>    * The :ref:`nrf_security_drivers_oberon` is currently limited to curve types secp224r1 and secp256r1 for ECDH and ECDSA.

Reviewer: What about Curve25519 ?

Author: Curve25519 is not possible for the algorithms ECDH or ECDSA. The corresponding equivalents for Curve25519 are x25519 (equivalent to ECDH) and Ed25519 (Also known as EdDSA, equivalent to ECDSA). There is no standardized APIs for this, yet. Only base curve operations are allowed for Curve25519
